### PR TITLE
Addtests, improve validation and error messaging

### PR DESCRIPTION
**Pull request by abergs**

### CODE REVIEW

File: src/Service/Fido2ServiceEndpoints.cs [hunk 1]
```diff
@@ -402,7 +410,15 @@ private void ValidateUserId(string userId)
...
-            throw new ApiException("userId must not be null or empty", 400);
+            throw new ApiException("Invalid UserId: UserId cannot be null or empty", 400);
+        }
+    }
+    
+    private void ValidateUsername(string username)
+    {
+        if (string.IsNullOrWhiteSpace(username))
+        {
+            throw new ApiException("Invalid Username: Username cannot be null or empty", 400);
```
Comment (justindbaur) [inline src/Service/Fido2ServiceEndpoints.cs:421]: Error code?

File: tests/Api.IntegrationTests/AuthorizationTests.cs [hunk 2]
```diff
@@ -59,9 +62,118 @@ public async Task ValidateThatEndpointsHaveProtection()
         }
     }
 
+    [Fact]
+    public async Task ValidateThatMissingApiSecretThrows()
+    {
+        var request = new HttpRequestMessage(HttpMethod.Get, "/credentials/list?userId=1");
+        request.Headers.Add("Accept", "application/json");
```
Comment (justindbaur) [inline tests/Api.IntegrationTests/AuthorizationTests.cs:69]: Should we add the `Accept` header ad a DefaultHeader for the entire client in the constructor, or alternatively even override `CreateClient` and do it in `TestWebApplicationFactory`?

File: tests/Api.IntegrationTests/AuthorizationTests.cs [hunk 3]
```diff
@@ -59,9 +62,118 @@ public async Task ValidateThatEndpointsHaveProtection()
...
+    public async Task ApiSecretGivesHelpfulAdvice(string input, string details)
+    {
+        var request = new HttpRequestMessage(HttpMethod.Get, "/credentials/list?userId=1");
+        request.Headers.Add("Accept", "application/json");
+
+        if (input == "public-header-instead")
+        {
+            request.Headers.Add("ApiKey", "something");
+        }
+        else if (input != "missing") {
```
Comment (justindbaur) [inline tests/Api.IntegrationTests/AuthorizationTests.cs:122]: Javascript style bracing 🥲 , you might need to run `dotnet format` at the root of the repo.

File: tests/Api.IntegrationTests/AuthorizationTests.cs [hunk 4]
```diff
@@ -59,9 +62,118 @@ public async Task ValidateThatEndpointsHaveProtection()
...
+    
+    [Theory]
+    [InlineData("test:secret:123", "Your ApiKey header contained a ApiSecret instead of your 'ApiKey'.")]
+    [InlineData("verify_123", "A verify token was supplied instead of your 'ApiKey'.")]
+    [InlineData("register_123", "A register token was supplied instead of your 'ApiKey'." )]
+    [InlineData("somethingrandom", "We don't recognize the value you supplied for your 'ApiKey'. It started with: 'somethingr'.")]
+    [InlineData("", "A valid 'ApiKey' header is required.")]
+    [InlineData("missing", "A valid 'ApiKey' header is required.")]
+    [InlineData("secret-header-instead", "A 'ApiSecret' header was supplied when a 'ApiKey' header should have been supplied.")]
+    public async Task ApiPublicGivesHelpfulAdvice(string input, string details)
```
Comment (justindbaur) [inline tests/Api.IntegrationTests/AuthorizationTests.cs:148]: Very nice tests, this one and the above one.

File: tests/Api.IntegrationTests/Helper.cs [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+            if (aCanMove == false && bCanMove == false)
+            {
+                // They both can't continue to enumerate at the same time, that is valid
+                break;
+            }
+
+            var aElement = aArrayEnumerator.Current;
+            var bElement = bArrayEnumerator.Current;
+
+            // Recursion!
```
Comment (justindbaur) [inline tests/Api.IntegrationTests/Helper.cs:229]: Recursion!

File: tests/Api.IntegrationTests/BasicTests.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        _client.DefaultRequestHeaders.Add("ApiSecret", _factory.ApiSecret);
+    }
+
+    public Task<HttpResponseMessage> PostAsync(string url, object payload)
+    {
+        var request = new HttpRequestMessage(HttpMethod.Post, url);
+        request.Content = JsonContent.Create(payload);
+        var httpResponse = _client.SendAsync(request);
+
+        return httpResponse;
```
Comment (justindbaur) [inline tests/Api.IntegrationTests/BasicTests.cs:25]:
```suggestion
        return await _client.PostAsJsonAsync(url, payload);
```

File: tests/Api.IntegrationTests/TestWebApplicationFactory.cs [hunk 7]
```diff
@@ -21,15 +25,23 @@ protected override IHost CreateHost(IHostBuilder builder)
...
+            var context = CreateDbContext(scope, "test");
             context.Database.EnsureDeleted();
             context.Database.EnsureCreated();
+            
+            // seed with test keys
+            // hash of secret
+            context.ApiKeys.Add(new ApiKeyDesc() { Tenant = "test", Id = "9795", ApiKey = "test:public:2e728aa5986f4ba8b073a5b28a939795" });
+            context.ApiKeys.Add(new ApiKeyDesc() { Tenant = "test", Id = "6d02", ApiKey = "4RtmMr0hVknaQAIhaRtPHw==:xR7bg3NVsC80a8GDDhH39g==",Scopes = new[] { "token_register", "token_verify" }, });
+            context.AccountInfo.Add(new AccountMetaInformation() { Tenant = "test", AcountName = "test" });
+            context.SaveChanges();
```
Comment (justindbaur) [inline tests/Api.IntegrationTests/TestWebApplicationFactory.cs:44]: Very nice!

File: src/Service/Fido2ServiceEndpoints.cs [hunk 8]
```diff
@@ -402,7 +410,15 @@ private void ValidateUserId(string userId)
     {
         if (string.IsNullOrWhiteSpace(userId))
         {
-            throw new ApiException("userId must not be null or empty", 400);
+            throw new ApiException("Invalid UserId: UserId cannot be null or empty", 400);
```
Comment (justindbaur) [inline src/Service/Fido2ServiceEndpoints.cs:413]: Error code?

**Review by justindbaur — CHANGES_REQUESTED** [review]

File: tests/Api.IntegrationTests/AuthorizationTests.cs (hunk 2 above)
Comment (abergs) [reply tests/Api.IntegrationTests/AuthorizationTests.cs:69]: Yeah, I think we can add it to `TestWebApplicationFactory `.

But I'm also a little bit worried -- I think we should default the entire API to have accept:application/json unless the caller sends us anything different.

**Review by justindbaur — APPROVED** [review]

**Pull request merged**